runtime/benchmark_test.go

```diff
@@ -0,0 +1,119 @@
+// Copyright 2018-2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package main
+
+import (
+	"context"
+	"testing"
+
+	"github.com/containerd/containerd"
+	"github.com/containerd/containerd/namespaces"
+	"github.com/containerd/containerd/pkg/ttrpcutil"
+	"github.com/firecracker-microvm/firecracker-containerd/proto"
+	fccontrol "github.com/firecracker-microvm/firecracker-containerd/proto/service/fccontrol/ttrpc"
+	"github.com/gofrs/uuid"
+	"github.com/stretchr/testify/require"
+)
+
+func createAndStopVM(
+	ctx context.Context,
+	fcClient fccontrol.FirecrackerService,
+	request proto.CreateVMRequest,
+) error {
+	uuid, err := uuid.NewV4()
+	if err != nil {
+		return err
+	}
+
+	request.VMID = uuid.String()
+
+	_, err = fcClient.CreateVM(ctx, &request)
+	if err != nil {
+		return err
+	}
+
+	_, err = fcClient.StopVM(ctx, &proto.StopVMRequest{VMID: request.VMID})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func benchmarkCreateAndStopVM(b *testing.B, vcpuCount uint32, kernelArgs string, parallel int) {
+	require := require.New(b)
+
+	client, err := containerd.New(containerdSockPath, containerd.WithDefaultRuntime(firecrackerRuntime))
+	require.NoError(err, "unable to create client to containerd service at %s, is containerd running?", containerdSockPath)
+	defer client.Close()
+
+	ctx := namespaces.WithNamespace(context.Background(), "default")
+
+	pluginClient, err := ttrpcutil.NewClient(containerdSockPath + ".ttrpc")
+	require.NoError(err, "failed to create ttrpc client")
+
+	fcClient := fccontrol.NewFirecrackerClient(pluginClient.Client())
+	request := proto.CreateVMRequest{
+		KernelArgs: kernelArgs,
+		MachineCfg: &proto.FirecrackerMachineConfiguration{
+			VcpuCount: vcpuCount,
+		},
+	}
+
+	ch := make(chan error, parallel)
+	for i := 0; i < parallel; i++ {
+		go func() {
+			defer func() {
+				ch <- nil
+			}()
+
+			for i := 0; i < b.N; i++ {
+				err := createAndStopVM(ctx, fcClient, request)
+				if err != nil {
+					ch <- err
+					return
+				}
+			}
+		}()
+	}
+
+	// Make sure all goroutines are finished successfully.
+	for i := 0; i < parallel; i++ {
+		require.NoError(<-ch)
```

Comment:
If there are errors we will write more than parallel times to ch I think? Worst case is every CreateAndStopVM() call errors which would be (b.N+1)*parallel iiuc. I'm not sure if that matters here as we are failing on *any* error though. Can you confirm?

Reply:
I think all goroutines will write either 1 (because of `defer`) or 2 (because of `ch <- err` and `defer`). Yes. That is wrong.

Reply:
Addressed in the last revision.